app/controllers/api/authentications_controller.rb

```diff
@@ -2,21 +2,25 @@ module Api
   class AuthenticationsController < BaseController
     def edit_resource(type, id, data)
       api_resource(type, id, "Updating") do |auth|
-        ensure_respond_to(type, auth, :update, :update_in_provider_queue)
+        ensure_supports(type, auth, :update)
         {:task_id => auth.update_in_provider_queue(data.deep_symbolize_keys)}
       end
     end
 
-    def create_resource(_type, _id, data)
+    def create_resource(type, _id, data)
       manager_resource, attrs = validate_auth_attrs(data)
-      task_id = AuthenticationService.create_authentication_task(manager_resource, attrs)
+
+      klass = ::Authentication.descendant_get(attrs['type'])
+      ensure_supports(type, klass, :create)
```

Comment:
Darn, this is very close to the `api_resource` model.

almost like we want to define:
```ruby
def resource_search(id, type)
  ::Authentication.descendant_get(type)
end

def create_resource(type, _id, data)
  api_resource(type, nil, "Creating", :supports => :create) do |klass|
    manager_resource, attrs = validate_auth_attrs(data)
    {:task_id =} klass.create_in_provider_queue(manager_resource.id, attrs.deep_symbolize_keys)}
  end
end
```

But may need to fall back to `api_action`.

Now that I think about it, I had refactored this method a number of times.
Never got it quite right.

---

This is a definite improvement.

Reply:
ooh, right

I kept splitting apart the `validate_auth_attrs`:


```ruby

def create_resource(_type, _id, data)
  type, id = data_resource(data)
  # maybe attrs from data_resource ?
  api_resource(type, id, "Creating", :supports => :create) do |auth, klass|
    {:task_id => klass.create_in_provider_queue(auth.id, attrs.deep_symbolize_keys)}
  end
end

def data_resource(data)
  href = Href.new(data['manager_resource']['href'])
  raise 'invalid manager_resource href specified' unless href.subject && href.subject_id
  [href.subject, href.subject_id]
end
```

nvr mind. there is a reason I never created a PR for this.